Replace for-loops with `foreach` in dmodule

dkorpel opened the pull request.

dkorpel committed:
Replace for-loops with foreach in dmodule

UplinkCoder commented:
@dkorpel double check that no loops you are doing can expand the set of things you are looping over.

dkorpel commented:
> @dkorpel double check that no loops you are doing can expand the set of things you are looping over.

I noticed that `runDeferredSemantic2` and `runDeferredSemantic3` added to the imports it was looping over, so I changed that back and added a comment. The rest looks okay to me, but I appreciate a second opinion.

UplinkCoder commented:
@dkorpel I'll have a look later.

WalterBright commented on src/dmd/dmodule.d:
```diff
@@ -1237,11 +1236,8 @@ extern (C++) final class Module : Package
-            for (size_t i = 0; i < members.dim; i++)
-            {
-                Dsymbol s = (*members)[i];
+            foreach (Dsymbol s; *members)
```
Do not replace looks over members with foreach. The reason is because many routines, as part of their work, add members to `members` and foreach assumes that the loop length does not change.

dkorpel replied:
Thanks, I reverted those changes. I noticed that the for-loop over members is a very common pattern in dmd, so I don't want to make changes to it only in dmodule. I might try to add an `InputRange` to iterate over Dsymbol members in a future PR, because currently it's not at all clear whether these for-loops are an archaism or intended.

MoonlightSentinel replied:
> Do not replace looks over members with foreach. The reason is because many routines, as part of their work, add members to `members` and foreach assumes that the loop length does not change.

Maybe we should add an `opApply` to `Array` which does the right thing by default?

dkorpel replied:
An array allows arbitrary insertion and removal, so there's no single right way to iterate over an array while modifying it during iteration.

MoonlightSentinel replied:
True, but those situations would require an index-based loop anyways. `opApply` would be an improvement over the current `opSlice` based iteration  because it does the right thing in the most common cases (no change or appending).

WalterBright commented on src/dmd/dmodule.d:
```diff
@@ -1250,14 +1246,12 @@ extern (C++) final class Module : Package
          */
         setScope(sc); // remember module scope for semantic
-        for (size_t i = 0; i < members.dim; i++)
+        foreach (Dsymbol s; *members)
```
ditto

WalterBright commented on src/dmd/dmodule.d:
```diff
@@ -1250,14 +1246,12 @@ extern (C++) final class Module : Package
             s.setScope(sc);
         }
-        for (size_t i = 0; i < members.dim; i++)
+        foreach (Dsymbol s; *members)
```
ditto

WalterBright commented on src/dmd/dmodule.d:
```diff
@@ -1442,6 +1436,7 @@ extern (C++) final class Module : Package
         Module.runDeferredSemantic();
 
         Dsymbols* a = &Module.deferred2;
+        // Note: a.dim can change inside the loop-body
```
What I am referring too :-) but sometimes it is not so obvious.

thewilsonator commented on src/dmd/dmodule.d:
```diff
@@ -1473,11 +1472,8 @@ extern (C++) final class Module : Package
-        for (size_t i = 0; i < amodules.dim; i++)
-        {
-            Module m = amodules[i];
+        foreach (Module m; amodules)
```
`ref`?

dkorpel replied:
I tried to keep the code equivalent, even though it doesn't matter since `Module` is a class. I can change them all to `ref` if that's preferred.

RazvanN7 commented on src/dmd/dmodule.d:
```diff
@@ -481,11 +481,11 @@ extern (C++) final class Module : Package
         {
-            for (size_t i = 0; i < amodules.dim; i++)
-                amodules[i].insearch = 0;
+            foreach (ref m; amodules)
```
Isn't this `ref` superfluous?

RazvanN7 commented on src/dmd/dmodule.d:
```diff
@@ -1241,7 +1240,7 @@ extern (C++) final class Module : Package
                 Dsymbol s = (*members)[i];
                 s.addMember(sc, sc.scopesym);
-            }
+        }
```
This change isn't right. Probably a left-over from previous versions.

RazvanN7 reviewed (dismissed):
Some `foreach`s use `(ref m; amodules)` whereas others use `(Module m; amodules)`. I suggest choosing one and using it consistently.  The latter seems better to me.

thewilsonator commented:
CircleCI seems stuck, might want to rebase or force push.

ibuclaw commented on src/dmd/dmodule.d:
```diff
@@ -1422,7 +1421,7 @@ extern (C++) final class Module : Package
             deferred.setDim(0);
 
-            for (size_t i = 0; i < len; i++)
+            foreach (i; 0..len)
```
This loop calls dsymbolSemantic, so better leave it as a `for`.

dkorpel replied:
len is a local variable in this function, dsymbolSemantic can't possibly change that.

MoonlightSentinel commented:
`CircleCI` seems stuck, maybe rebase to restart the CI?
